operator/internal/controller/charactercounter_controller.go

```diff
@@ -47,16 +55,138 @@ type CharacterCounterReconciler struct {
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.15.0/pkg/reconcile
 func (r *CharacterCounterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	r.log = log.FromContext(ctx)
+	r.log.Info("start reconcile for", "name", req.Name)
 
-	// TODO(user): your logic here
+	var cc rampupv1alpha1.CharacterCounter
+	if err := r.Get(ctx, req.NamespacedName, &cc); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(fmt.Errorf("unable to fetch ChracterCounter CRD: %w", err))
+	}
 
-	return ctrl.Result{}, nil
+	if cc.Spec.Namespace == "" {
+		cc.Spec.Namespace = "default"
+	}
+
+	r.cc = &cc
+
+	r.log.Info("server reconcileService")
+	if err := r.reconcileService(ctx, cc.Spec.Server); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	r.log.Info("server reconcileDeployment")
+	if err := r.reconcileDeployement(ctx, cc.Spec.Server, false); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	r.log.Info("frontend reconcileService")
+	if err := r.reconcileService(ctx, cc.Spec.Frontend); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	r.log.Info("frontend reconcileDeployment")
+	return ctrl.Result{}, r.reconcileDeployement(ctx, cc.Spec.Frontend, true)
+}
+
+func (r *CharacterCounterReconciler) reconcileService(ctx context.Context, component rampupv1alpha1.CharacterCounterComponent) error {
+	var service corev1.Service
+	service.Name = component.Name
+	service.Namespace = r.cc.Namespace
+	service.Labels = r.cc.Spec.Labels
+	op, err := ctrl.CreateOrUpdate(ctx, r.Client, &service, func() error {
+		modifyService(*r.cc, component, &service)
+		return ctrl.SetControllerReference(r.cc, &service, r.Scheme)
+	})
+	if err != nil {
+		return fmt.Errorf("could not create or update service: %w", err)
+	}
+	if op != controllerutil.OperationResultNone {
+		r.log.Info("reconcile service successfully", "operation", op)
+	}
+	return nil
+}
+
+func modifyService(cc rampupv1alpha1.CharacterCounter, component rampupv1alpha1.CharacterCounterComponent, service *corev1.Service) {
+	service.Labels = cc.Spec.Labels
+	service.Spec = corev1.ServiceSpec{
+		Ports:    component.ServicePorts,
+		Selector: component.Selector,
+	}
+}
+
+func (r *CharacterCounterReconciler) reconcileDeployement(ctx context.Context, component rampupv1alpha1.CharacterCounterComponent, frontend bool) error {
+	args := []string{}
+	if frontend {
+		args = append(args, "-grpc-addr", fmt.Sprintf("%s:80", r.cc.Spec.Server.Name))
+	}
+	deployment := appsv1.Deployment{
+		TypeMeta: metav1.TypeMeta{APIVersion: appsv1.SchemeGroupVersion.String(), Kind: "Deployment"},
```

Comment:
This is only needed in very rare scenarios. This is not one of them

Reply:
by `this` you mean the TypeMeta?

Reply:
yes

Reply:
removed